command/debug/debug.go

```diff
@@ -320,212 +323,217 @@ func (c *cmd) captureStatic() error {
 		marshaled, err := json.MarshalIndent(v, "", "\t")
 		if err != nil {
-			errors = multierror.Append(errors, err)
+			errs = multierror.Append(errs, err)
 		}
 
 		err = ioutil.WriteFile(fmt.Sprintf("%s/%s.json", c.output, output), marshaled, 0644)
 		if err != nil {
-			errors = multierror.Append(errors, err)
+			errs = multierror.Append(errs, err)
 		}
 	}
 
-	return errors
+	return errs
 }
 
-// captureDynamic blocks for the duration of the command
+// captureInterval blocks for the duration of the command
 // specified by the duration flag, capturing the dynamic
 // targets at the interval specified
-func (c *cmd) captureDynamic() error {
-	successChan := make(chan int64)
-	errCh := make(chan error)
+func (c *cmd) captureInterval() error {
+	intervalChn := time.NewTicker(c.interval)
+	defer intervalChn.Stop()
 	durationChn := time.After(c.duration)
 	intervalCount := 0
 
 	c.UI.Output(fmt.Sprintf("Beginning capture interval %s (%d)", time.Now().Local().String(), intervalCount))
 
-	// We'll wait for all of the targets configured to be
-	// captured before continuing
-	var wg sync.WaitGroup
+	err := captureShortLived(c)
+	if err != nil {
+		return err
+	}
+	for {
+		select {
+		case t := <-intervalChn.C:
+			intervalCount++
+			c.UI.Output(fmt.Sprintf("Capture successful %s (%d)", t.Local().String(), intervalCount))
+			err := captureShortLived(c)
+			if err != nil {
+				return err
+			}
+		case <-durationChn:
+			intervalChn.Stop()
+			return nil
+		case <-c.shutdownCh:
+			return errors.New("stopping collection due to shutdown signal")
+		}
+	}
+}
 
-	capture := func() {
-		timestamp := time.Now().Local().Unix()
+func captureShortLived(c *cmd) error {
+	g := new(errgroup.Group)
+	timestamp := time.Now().Local().Unix()
 
-		// Make the directory that will store all captured data
-		// for this interval
-		timestampDir := fmt.Sprintf("%s/%d", c.output, timestamp)
-		err := os.MkdirAll(timestampDir, 0755)
-		if err != nil {
-			errCh <- err
-		}
+	timestampDir, err := c.createTimestampDir(timestamp)
+	if err != nil {
+		return err
+	}
+	if c.configuredTarget("pprof") {
+		g.Go(func() error {
+			return c.captureHeap(timestampDir)
+		})
 
-		// Capture metrics
-		if c.configuredTarget("metrics") {
-			wg.Add(1)
-
-			go func() {
-				metrics, err := c.client.Agent().Metrics()
-				if err != nil {
-					errCh <- err
-				}
-
-				marshaled, err := json.MarshalIndent(metrics, "", "\t")
-				if err != nil {
-					errCh <- err
-				}
-
-				err = ioutil.WriteFile(fmt.Sprintf("%s/%s.json", timestampDir, "metrics"), marshaled, 0644)
-				if err != nil {
-					errCh <- err
-				}
-
-				// We need to sleep for the configured interval in the case
-				// of metrics being the only target captured. When it is,
-				// the waitgroup would return on Wait() and repeat without
-				// waiting for the interval.
-				time.Sleep(c.interval)
-
-				wg.Done()
-			}()
-		}
+		g.Go(func() error {
+			return c.captureGoRoutines(timestampDir)
+		})
```

Comment:
🤔 is there any possible downside to capturing these in parallel? for example the heap dump might show memory usage that was in fact only due to the concurrent goroutine dump being run. Both have some slightl overhead too so doing both at the same time potentially compounds the overhead on a production server... 

Honestly, I doubt either is a problem in practice but just made me wonder. Would it be safer just to do each in series here?

Reply:
Hmm looks like we used to go profile and trace in parallel with goroutines anyway and then heap separately. This probably is fine as it is!

Reply:
previously goroutine and heap were run in series (each alone). Profile and traces in parallel. I can extract heap and profile in a separate routine that run before we run profile and trace captures and then run profile and trace in parallel with metrics and logs

Reply:
I investigated the changes needed for this one and it's hard to separate because we run the profile and trace for the whole duration without interruption now we can't stop them while doing the goroutine and heap dump. the other option I can think of is to introduce another level of granularity where we can specify the pprof metrics we want to capture any of (cpu, heap, goroutine and trace). We should do it in a separate PR, this one have already a lot of changes.

Reply:
Yeah I came to the conclusion it probably doesn't matter anyway - unless we see issues with this in practice I'd say leave it like this for now 👍